staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go

```diff
@@ -0,0 +1,250 @@
+	if p.rw.WasHijacked() || p.rw.StatusCode() != 0 {
+		return false
+	}
+
+	if p.delegate.retry() {
+		p.delegate.reset()
+		return true
+	}
+	return false
+}
+
+type maxRetries struct {
+	retriable
+	counter int
+	max int
+}
+
+var _ retriable = &maxRetries{}
+
+func withMaxRetries(delegate retriable, max int) retriable {
+	return &maxRetries{retriable:delegate, max:max}
+}
+
+func (r *maxRetries) retry() bool {
+	r.counter++
+	if r.counter > r.max {
+		return false
+	}
+
+	return r.retriable.retry()
+}
+
+type backOff struct {
+	retriable
+}
+
+var _ retriable = &backOff{}
+
+func withBackOff(delegate retriable) retriable {
+	return &backOff{delegate}
+}
+
+func (b *backOff) retry() bool {
+	// TODO: implement back off
+	return b.retriable.retry()
+}
+
+// retriableHijackErrorResponder wraps proxy.ErrorResponder and prevents errors from being written to the client if they can be retried
+type retriableHijackErrorResponder struct {
+	// delegate knows how to write errors to the client
+	delegate proxy.ErrorResponder
+	req *http.Request
+	lastKnownError error
+	isRetriable func(req *http.Request, err error) bool
+}
+
+var _ proxy.ErrorResponder = &retriableHijackErrorResponder{}
+var _ retriable = &retriableHijackErrorResponder{}
+
+// newHijackErrorResponder creates a new ErrorResponder that wraps the delegate for supporting reties
+func newHijackErrorResponder(delegate proxy.ErrorResponder, req *http.Request) *retriableHijackErrorResponder {
+	return &retriableHijackErrorResponder{delegate: delegate, req: req}
+}
+
+// Error reports the err to the client or suppress it if it's not retriable
+func (hr *retriableHijackErrorResponder) Error(w http.ResponseWriter, r *http.Request, err error) {
+	// if we can retry the request do not send a response to the client
+	hr.lastKnownError = err
+	if !hr.isRetriable(hr.req, err) {
+		hr.delegate.Error(w, r, err) // this might send a response to a client
+	}
+}
+
+func (hr *retriableHijackErrorResponder) reset() {
+	hr.lastKnownError = nil
+}
+
+func (hr *retriableHijackErrorResponder) retry() bool {
+	return hr.isRetriable(hr.req, hr.lastKnownError)
+}
+
+// TODO: doc
+func withHijackErrorResponderForSingleEndpoint(hr *retriableHijackErrorResponder) retriable {
+	hr.isRetriable = func(req *http.Request, err error) bool {
+		if isHTTPVerbRetriable(req) && (knet.IsConnectionReset(err) || knet.IsProbableEOF(err) ||  isExperimental(err)) {
+			return true
+		}
+		return false
+	}
+	return hr
+}
+
+// TODO: doc
+func withHijackErrorResponderForMultipleEndpoints(hr *retriableHijackErrorResponder) retriable {
+	hr.isRetriable = func(req *http.Request, err error) bool {
+		if isHTTPVerbRetriable(req) && (knet.IsConnectionReset(err) || knet.IsConnectionRefused(err) || knet.IsProbableEOF(err) ||  isExperimental(err)) {
+			return true
+		}
+		return false
+	}
+	return hr
+}
+
+func isHTTPVerbRetriable(req *http.Request) bool {
+	return req.Method == "GET" || req.Method == "DELETE"
```

Comment:
for delete it depends on whether the resource version is given in DeleteOptions afaik.

Reply:
`DeleteOptions ` doesn't support `ResourceVersion`, I think only GET, LIST and WATCH do.